File: csv-query/typescript/src/CsvParser.ts
import { CsvTable } from './CsvTable.js';
import { Row } from './Row.js';

export function parseCsv(csv: string): CsvTable {
  const lines = csv.split('\n').filter(l => l.length > 0);
  if (lines.length === 0) return new CsvTable([], []);

  const headers = parseFields(lines[0]!);
  const rows: Row[] = [];

  for (let i = 1; i < lines.length; i++) {
    const fields = parseFields(lines[i]!);
    const values = new Map<string, string>();
    for (let j = 0; j < headers.length; j++) {
      values.set(headers[j]!, j < fields.length ? fields[j]! : '');
    }
    rows.push(new Row(values));
  }

  return new CsvTable(headers, rows);
}

function parseFields(line: string): string[] {
  const fields: string[] = [];
  let current = '';
  let inQuotes = false;

  for (const ch of line) {
    if (ch === '"') {
      inQuotes = !inQuotes;
    } else if (ch === ',' && !inQuotes) {
      fields.push(current.trim());
      current = '';
    } else {
      current += ch;
    }
  }

  fields.push(current.trim());
  return fields;
}

File: csv-query/csharp/src/CsvQuery/CsvParser.cs
namespace CsvQuery;

public static class CsvParser
{
    public static CsvTable Parse(string csv)
    {
        var lines = csv.Split('\n', StringSplitOptions.RemoveEmptyEntries);
        if (lines.Length == 0)
            return new CsvTable(Array.Empty<string>(), new List<Row>());

        var headers = ParseFields(lines[0]);
        var rows = new List<Row>();

        for (var i = 1; i < lines.Length; i++)
        {
            var fields = ParseFields(lines[i]);
            var values = new Dictionary<string, string>();
            for (var j = 0; j < headers.Length; j++)
            {
                values[headers[j]] = j < fields.Length ? fields[j] : "";
            }
            rows.Add(new Row(values));
        }

        return new CsvTable(headers, rows);
    }

    private static string[] ParseFields(string line)
    {
        var fields = new List<string>();
        var current = "";
        var inQuotes = false;

        for (var i = 0; i < line.Length; i++)
        {
            var ch = line[i];
            if (ch == '"')
            {
                inQuotes = !inQuotes;
            }
            else if (ch == ',' && !inQuotes)
            {
                fields.Add(current.Trim());
                current = "";
            }
            else
            {
                current += ch;
            }
        }

        fields.Add(current.Trim());
        return fields.ToArray();
    }
}
